### PR TITLE
Cache (some) existence checks during evaluation

### PR DESCRIPTION
**Pull request by cdmihai**
Some existence checks still remain:
- checks in sdk resolver discovery and sdk resolution. Ok not to cache because it happens once per sdk service, and the sdk service lifetime can be extended by the evaluation context. Also, sdk resolutions are cached as well. 
- checks in ToolLocationHelper and friends via property functions. Have to investigate whether those checks happen once per process or not. If not, I'll have to change the property function invoker to pass on the context when the target function can accept evaluation contexts
- checks in Intrinsic functions, like getting the file above. Same fix as the ToolLocationHelper required.

Introduced a new `MSBuildOnWindowsFileSystem` which does a mix of native and managed checks. Ideally, eventually, after further investigation and profiling, we can decide whether to have pure native implementations for each OS, or whether managed is good enough (at least for now, full framework allocates way too much memory, even for existence checks, as opposed to "0" allocations for the native calls).

### CODE REVIEW

**Comment by cdmihai** [comment]
Projects\msbuild is master, and projects\msbuild_3 is this branch

Project | projects\msbuild 1st run | projects\msbuild 2nd run | Improvement | projects\msbuild_3 1st run | projects\msbuild_3 2nd run | Improvement | Improvement diff
-- | -- | -- | -- | -- | -- | -- | --
DotnetConsoleProject | 49.7344 | 39.0155 | 21.55228574 | 48.8086 | 36.6164 | 24.97961425 | 3.427328506
DotnetWebProject | 63.443 | 55.0911 | 13.1644153 | 61.8 | 52.6194 | 14.85533981 | 1.690924504
DotnetMvcProject | 71.3054 | 57.7002 | 19.08018187 | 70.1563 | 55.9806 | 20.20588315 | 1.125701284
Picasso | 376.0456 | 385.4723 | -2.506797048 | 325.9202 | 329.1034 | -0.976680795 | 1.530116253
SmallP2POldCsproj | 67.497 | 68.4856 | -1.464657688 | 61.1354 | 59.5268 | 2.63120876 | 4.095866448
SmallP2PNewCsproj | 233.7214 | 205.2185 | 12.19524613 | 213.2571 | 186.815 | 12.39916514 | 0.203919004
LargeP2POldCsproj | 1112.8226 | 1117.8374 | -0.450637864 | 954.3222 | 901.5532 | 5.529474217 | 5.980112082
Generated_100_100_v150 | 1286.5278 | 1288.2522 | -0.134035191 | 999.8952 | 977.9322 | 2.196530196 | 2.330565388
OrchardCore | 3491.5815 | 2118.0614 | 39.33805068 | 3261.9078 | 1931.7956 | 40.77712436 | 1.439073677
Roslyn | 4194.858 | 3580.2368 | 14.65177606 | 3867.4398 | 3266.4183 | 15.54055218 | 0.888776121
WebLargeCore | 3905.276 | 2565.4188 | 34.30889904 | 3613.009 | 2332.6554 | 35.43732108 | 1.128422042

### SerialEvaluationSharedContextSecondRun Time (ms) (projects\msbuild vs projects\msbuild_3)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 39.0692 -> 36.7762 (-5.869%)
DotnetWebProject | :white_check_mark: | yes | 55.1948 -> 52.7017 (-4.517%)
DotnetMvcProject | :white_check_mark: | yes | 57.8913 -> 56.1712 (-2.971%)
Picasso | :white_check_mark: | yes | 388.5685 -> 331.2147 (-14.76%)
SmallP2POldCsproj | :white_check_mark: | yes | 68.9644 -> 59.7735 (-13.327%)
SmallP2PNewCsproj | :white_check_mark: | yes | 205.3771 -> 186.7665 (-9.062%)
LargeP2POldCsproj | :white_check_mark: | yes | 1116.6842 -> 909.4412 (-18.559%)
Generated_100_100_v150 | :white_check_mark: | yes | 1293.6938 -> 980.5454 (-24.206%)
OrchardCore | :white_check_mark: | yes | 2131.9463 -> 1931.1276 (-9.42%)
Roslyn | :white_check_mark: | yes | 3587.6697 -> 3278.1968 (-8.626%)
WebLargeCore | :white_check_mark: | yes | 2561.9186 -> 2330.4902 (-9.033%)

### SerialEvaluationSharedContextSecondRun Memory (bytes) (projects\msbuild vs projects\msbuild_3)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 6224048 -> 6235781 (0.189%)
Picasso | :white_check_mark: | yes | 11605978 -> 11055720 (-4.741%)
SmallP2PNewCsproj | :white_check_mark: | yes | 11622924 -> 11344553 (-2.395%)
LargeP2POldCsproj | :red_circle: | yes | 22733928 -> 24485666 (7.705%)
Generated_100_100_v150 | :white_check_mark: | yes | 28564908 -> 26199969 (-8.279%)
OrchardCore | :ok_hand: | no | 55983853 -> 56319904 (0.6%)
Roslyn | ::ok_hand: | no | 85235553 -> 85077057 (-0.186%)
WebLargeCore | :ok_hand: | no | 65143259 -> 65421763 (0.428%)

### SerialEvaluationSharedContext Time (ms) (projects\msbuild vs projects\msbuild_3)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 49.8277 -> 48.9455 (-1.771%)
DotnetWebProject | :white_check_mark: | yes | 63.7792 -> 61.8116 (-3.085%)
DotnetMvcProject | :white_check_mark: | yes | 71.5251 -> 70.4611 (-1.488%)
Picasso | :white_check_mark: | yes | 377.0424 -> 327.5398 (-13.129%)
SmallP2POldCsproj | :white_check_mark: | yes | 67.9203 -> 61.1734 (-9.934%)
SmallP2PNewCsproj | :white_check_mark: | yes | 234.9273 -> 213.765 (-9.008%)
LargeP2POldCsproj | :white_check_mark: | yes | 1117.3472 -> 948.643 (-15.099%)
Generated_100_100_v150 | :white_check_mark: | yes | 1288.6539 -> 1002.1606 (-22.232%)
OrchardCore | :white_check_mark: | yes | 3485.6572 -> 3259.1676 (-6.498%)
Roslyn | :white_check_mark: | yes | 4205.6105 -> 3876.7053 (-7.821%)
WebLargeCore | :white_check_mark: | yes | 3892.9926 -> 3613.4676 (-7.18%)

### SerialEvaluationSharedContext Memory (bytes) (projects\msbuild vs projects\msbuild_3)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 3636447 -> 3644204 (0.213%)
DotnetWebProject | :red_circle: | yes | 5237922 -> 5251982 (0.268%)
DotnetMvcProject | :red_circle: | yes | 6027442 -> 6045660 (0.302%)
Picasso | :red_circle: | yes | 11618669 -> 11932274 (2.699%)
SmallP2POldCsproj | :red_circle: | yes | 4736150 -> 4779174 (0.908%)
SmallP2PNewCsproj | :red_circle: | yes | 7630246 -> 8589090 (12.566%)
LargeP2POldCsproj | :red_circle: | yes | 23569136 -> 25370432 (7.643%)
Generated_100_100_v150 | :red_circle: | yes | 26097569 -> 27085007 (3.784%)
OrchardCore | :red_circle: | yes | 59593851 -> 60520968 (1.556%)
Roslyn | :red_circle: | yes | 87051586 -> 87643569 (0.68%)
WebLargeCore | :ok_hand: | no | 69825760 -> 71194489 (1.96%)

**Comment by cdmihai** [comment]
Largest improvements seem to be for the old style csprojes and c++ (Generated_100_100_v150)

**Comment by cdmihai** [comment]
@dotnet-bot test OSX10.13 Build for CoreCLR please

File: src/FileSystemSources.props [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<Project>
```
Comment (AndyGerlicher) [inline src/FileSystemSources.props]: I like this! But maybe put it under shared? And `.props` doesn't seem like the right extension. Maybe `FileSystemSources.proj`? But I really do like the model over what we have today.

File: src/Shared/FileUtilities.cs [hunk 2]
```diff
@@ -824,32 +827,30 @@ internal static FileInfo GetFileInfoNoThrow(string filePath)
         /// </summary>
         /// <param name="fullPath">Full path to the directory in the filesystem</param>
         /// <returns></returns>
-        internal static bool DirectoryExistsNoThrow(string fullPath)
+        internal static bool DirectoryExistsNoThrow(string fullPath
+#if !CLR2COMPATIBILITY
```
Comment (AndyGerlicher) [inline src/Shared/FileUtilities.cs]: Would it streamline things to make a CLR2COMPATIBILITY FileSystem that was as basic as possible? We don't care about perf _at all_ in that scenario just about code cleanliness. But if it's more trouble than it's worth this is fine.

File: src/Shared/NativeMethodsShared.cs [hunk 3]
```diff
@@ -1423,12 +1423,50 @@ internal static bool MsgWaitOne(this WaitHandle handle, int timeout)
...
+        }
+
+        internal static bool DirectoryExistsWindows(string fullPath)
+        {
+            NativeMethodsShared.WIN32_FILE_ATTRIBUTE_DATA data = new NativeMethodsShared.WIN32_FILE_ATTRIBUTE_DATA();
+            bool success = false;
+
+            success = NativeMethodsShared.GetFileAttributesEx(fullPath, 0, ref data);
+            return success && (data.fileAttributes & NativeMethodsShared.FILE_ATTRIBUTE_DIRECTORY) != 0;
+        }
```
Comment (AndyGerlicher) [inline src/Shared/NativeMethodsShared.cs]: nit: add new line here

**Comment by AndyGerlicher** [comment]
Improvement looks great!

**Pull request merged**